### PR TITLE
Introduce a `toParent()` variant that strictly follows the declaration context

### PR DESCRIPTION
**Pull request by SSoulaimane**
Fragment from #9282 

Not exceptionally useful yet as divergence between template parents is not very common aside from static member templates.

### CODE REVIEW

**Review by thewilsonator — APPROVED** [review]

**Comment by thewilsonator** [comment]
Thanks! We can come up with a more descriptive name more these later.

File: src/dmd/dsymbol.d [hunk 1]
```diff
@@ -476,6 +476,9 @@ extern (C++) class Dsymbol : ASTNode
      * `toParent3()` returns a logically enclosing scope symbol this is a member of.
      * It skips over TemplateMixin's.
      *
+     * `toParent4()` similar to `toParent2()` but always follows the template declaration scope
+     * instead of the instantiation scope.
```
Comment (SSoulaimane) [inline src/dmd/dsymbol.d]: I should make it clear that it ignores static templates

File: src/dmd/semantic3.d [hunk 2]
```diff
@@ -1074,7 +1074,7 @@ private extern(C++) final class Semantic3Visitor : Visitor
                 {
                     /* Wrap the entire function body in a synchronized statement
                      */
-                    ClassDeclaration cd = funcdecl.isThis() ? funcdecl.isThis().isClassDeclaration() : funcdecl.parent.isClassDeclaration();
+                    ClassDeclaration cd = funcdecl.toParent4().isClassDeclaration();
```
Comment (SSoulaimane) [inline src/dmd/semantic3.d]: I should fix this. This doesn't work with static member template functions. I should probably make it a separate pull request since it was already erroneous before this change.

The following peace of code shouldn't fail.
```d
class C
{
    static synchronized func(alias a)() {}
}

void main()
{
    int a;
    C.func!a();
}
```

File: src/dmd/dsymbol.d (hunk 1 above)
Comment (SSoulaimane) [reply src/dmd/dsymbol.d]: I'll probably change the function to suit what the comment says and rewrite some code to accommodate for that t makes ore sense this way.
